### PR TITLE
RDCC-2362: MFA - PUT API to modify the MFA preference of a particular organisation

### PR DESCRIPTION
**Pull request by ishratt**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDCC-2362

### Change description ###
PUT API to modify the MFA preference of a particular organisation

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalExternalUserFunctionalTest.java
Comment (AdilOoze) [inline src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalExternalUserFunctionalTest.java:439]: Does this method need to be called again if it is called on line 62?

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/client/ProfessionalApiClient.java
Comment (abhijitdiwan) [inline src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/client/ProfessionalApiClient.java:793]: put logging component name and also use {} place holder and replace +

File: src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/LaunchDarklyConfiguration.java
Comment (AdilOoze) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/LaunchDarklyConfiguration.java:35]: I may be wrong, but will it match for `{orgId}` and not an actual Organisation ID? Please test this before merging

File: src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/FindMFAByUserIDTest.java
Comment (abhijitdiwan) [inline src/integrationTest/java/uk/gov/hmcts/reform/professionalapi/FindMFAByUserIDTest.java:57]: with sonar it ask assert Map with entry. It is good to change not mandatory.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java
Comment (AdilOoze) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java:52]: I think we should not return `200` here but instead from the `mfaStatusService` method. Otherwise we will always return 200 success when it may not be successful - we had similar issue to another API before

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java
Comment (AdilOoze) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java:474]: I think we should not return `200` here but instead from the `mfaStatusService` method. Otherwise we will always return `200` success when it may not be successful - we had similar issue to another API before

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java:461]: please replace ... very minor though but we should be consistent & use ::

File: src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java
Comment (AdilOoze) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java:187]: I believe we are using `@Getter/@Setter` - is this required?

File: src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java
Comment (AdilOoze) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java:131]: I believe we are using `@Getter/@Setter` - is this required?

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalController.java:468]: can we use constant for exception message?

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java
Comment (AdilOoze) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java:20]: Do we require a new Controller class for this API? Or can we not place this is `OrganisationExternalController`?

File: src/test/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalControllerTest.java
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/professionalapi/controller/internal/OrganisationInternalControllerTest.java:384]: Please add a `verify()` for this `when()`

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalExternalUserFunctionalTest.java
Comment (ishratt) [reply src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalExternalUserFunctionalTest.java:439]: yes, because even though it is part of the same class, this is considered as a separate test as we have added the @Test annotation, so it needs to be run again.

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java
Comment (ishratt) [reply src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalInternalUserFunctionalTest.java:402]: yes, because even though it is part of the same class, this is considered as a separate test as we have added the @Test annotation, so it needs to be run again.

File: src/test/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImplTest.java
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImplTest.java:49]: Please add `verify()` for these `when()`

File: src/test/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImplTest.java
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImplTest.java:81]: Please add a `verify()` for this `when()`

**Review by AdilOoze — COMMENTED** [review]
Please ensure that there is sufficient/complete mutation coverage for all the new classes and methods you have written

File: src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalExternalUserFunctionalTest.java
Comment (AdilOoze) [reply src/functionalTest/java/uk/gov/hmcts/reform/professionalapi/ProfessionalExternalUserFunctionalTest.java:439]: Can it not be ran in the same test method as the other classes? Otherwise we will be creating extra organisations and users

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java
Comment (ishratt) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java:20]: Yes, @shreedhar-hmcts suggested we create a separate controller for this

File: src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java
Comment (ishratt) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java:187]: all the other fields in this entity have explicit getters, so just tried to stay consistent but can remove it if required.

Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/domain/Organisation.java:187]: I don't think it is required, remove and see if everything compiles fine - if not add back in

File: src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/SecurityConfiguration.java
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/SecurityConfiguration.java:62]: I believe we do not want any web security like User and Service tokens around this API. Please let me know if there is some other scenario.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/advice/ExceptionMapper.java
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/controller/advice/ExceptionMapper.java:185]: Can we please move this to a constant file.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/controller/external/OrganisationMfaStatusController.java:52]: Good catch Adil.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/SecurityConfiguration.java
Comment (ishratt) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/SecurityConfiguration.java:62]: that's right, this is only for MFA - GET api, the rest should always need authentication.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImpl.java
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImpl.java:45]: Before reaching this code line, I believe that we have a default MFA set for all existing organisations. It's better to add a null check just to avoid any edge case failures.
What should happen if the default MFA is not available for an organisation. I think BA can assess this and we can update the code accordingly.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImpl.java
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImpl.java:48]: Kindly put all Strings in a constant file so that we can reuse in test cases.

File: src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/SecurityConfiguration.java
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/configuration/SecurityConfiguration.java:62]: Thanks Ishrat.

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImpl.java
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/professionalapi/service/impl/MfaStatusServiceImpl.java:49]: As discussed, this scenario is not likely to occur in real world. Marking this conversation as resolved.

**Review by SujeethKumarAri — DISMISSED** [review]

**Review by nitishsingla1234 — APPROVED** [review]

**Review by kavithagade12 — DISMISSED** [review]

**Review by shreedhar-hmcts — CHANGES_REQUESTED** [review]
Just putting this block until we get approval for master merge .

**Review by kavithagade12 — APPROVED** [review]

**Review by shreedhar-hmcts — APPROVED** [review]

**Pull request merged**
